File: 1/ruby/first.rb
# frozen_string_literal: true

#
# P:
#
# We need to find the password from a series of rotations of a circular dial.
# The password is the number of times the dial is left pointing at 0 after
# and rotation in a given sequence of rotations of the dial. The dial loops around
# so that rotating the dial by 1 to the right from 99 gives you 0, etc.
# The dial starts at 50.
#
#
# Etc:
#
# DS:
#
# class Dial
# @number --- starts at 50, min 0 and max 99
# #number --- get @number
# #rotate(direction, clicks) --- rotates dial in the given direction by clicks
#
# class Parser
#
# A:
#

# The dial
class Dial
  attr_reader :number

  def initialize
    @number = 50
  end

  def turn!(direction, clicks)
    clicks = -clicks if direction == 'L'
    @number = (@number + clicks) % 100
  end
end

# Parse rotations given in input file string
class Parser
  VALID_DIRECTIONS = %w[L R].freeze
  class << self
    def process(input)
      result = []
      lines = input.split("\n")
      lines.each do |line|
        next unless VALID_DIRECTIONS.include?(line[0].upcase)

        rotation = {}
        rotation[:direction] = line[0].upcase
        rotation[:clicks] = line[1..].to_i

        result.push rotation
      end

      result
    end
  end
end

# Orchestration class
class FindPassword
  class << self
    def from(file)
      # rubocop:disable Style/ExpandPathArguments
      root = File.expand_path('..', __FILE__)
      # rubocop:enable Style/ExpandPathArguments
      input = File.read(File.join(root, file))
      rotations = Parser.process(input)
      dial = Dial.new
      acc = 0
      rotations.each do |rotation|
        dial.turn!(rotation[:direction], rotation[:clicks])
        acc += 1 if dial.number.zero?
      end
      puts acc
    end
  end
end

FindPassword.from 'input1.txt'

File: 1/cpp/first.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <vector>
#include <algorithm>

class FileError {};

struct Rotation
{
    char direction;
    int clicks;
};

class Dial
{
public:
    Dial() : number{ 50 } {}

    int getNumber() const;
    int turn(Rotation rotation);
private:
    int number;
};

int Dial::getNumber() const
{
    return number;
}

int Dial::turn(Rotation rotation)
{
    int clicks{ rotation.clicks };
    if (rotation.direction == 'L') {
        clicks = -clicks;
    }
    number = (number + clicks) % 100;
    return number;
}

std::vector<Rotation> parse(const std::string& filename)
{
    std::vector<Rotation> result{};
    std::ifstream file{ filename };
    if (!file) {
        throw FileError{};
    }
    std::string line{};
    while (std::getline(file, line)) {
        Rotation r;
        r.direction = line.front();
        r.clicks = std::stoi(line.substr(1, line.size() - 1));
        result.push_back(r);
    }
    return result;
}

int main()
try {
    using namespace std::string_literals;

    std::vector<Rotation> rotations{ parse("input1.txt"s) };
    Dial dial{};
    int acc{ 0 };
    std::for_each(rotations.cbegin(), rotations.cend(), [&acc, &dial](Rotation r) {
        dial.turn(r);
        if (dial.getNumber() == 0) {
            acc += 1;
        }
    });
    std::cout << acc << '\n';

}
catch (const FileError& e) {
    std::cerr << "Can't read from file.\n";
    return 1;
}
